system_test/aest_hard_fork_SUITE.erl

```diff
@@ -0,0 +1,564 @@
+    [{_, _} = proplists:lookup(db_backup_top_height, SavedCfg),
+     {_, _} = proplists:lookup(db_backup_top_hash, SavedCfg)
+     | Config];
+init_per_group(upgrade_flow_smoke_test, Config) ->
+    {_, TopHeight} = proplists:lookup(db_backup_top_height, Config),
+    NewConfig = aest_nodes:ct_setup(Config),
+    aest_nodes:setup_nodes(
+      [?NEW_NODE1(?HEIGHT_OF_NEW_PROTOCOL(TopHeight)),
+       ?NEW_NODE2(?HEIGHT_OF_NEW_PROTOCOL_FOR_VALIDATING_BLOCKS(TopHeight)),
+       ?NEW_NODE3(?HEIGHT_OF_NEW_PROTOCOL(TopHeight)),
+       ?NEW_NODE4(?HEIGHT_OF_NEW_PROTOCOL(TopHeight))], NewConfig),
+    NewConfig.
+
+end_per_group(assumptions, _) -> ok;
+end_per_group(genesis, _) -> ok;
+end_per_group(hard_fork, _) -> ok;
+end_per_group(hard_fork_all, _) -> ok;
+end_per_group(upgrade_flow_smoke_test, Config) ->
+    aest_nodes:ct_cleanup(Config),
+    ok.
+
+init_per_testcase(old_node_persisting_chain_and_not_mining_has_genesis_as_top, Config) ->
+    aest_nodes:ct_setup(Config);
+init_per_testcase(new_node_persisting_chain_and_not_mining_has_same_old_genesis_as_top, Config) ->
+    aest_nodes:ct_setup(Config);
+init_per_testcase(new_nodes_can_mine_and_sync_fast_minimal_chain_with_pow, Config) ->
+    aest_nodes:ct_setup(Config);
+init_per_testcase(restore_db_backup_on_old_node, Config) ->
+    aest_nodes:ct_setup(Config);
+init_per_testcase(old_node_can_receive_chain_from_other_old_node, Config) ->
+    aest_nodes:ct_setup(Config);
+init_per_testcase(restore_db_backup_with_short_chain_on_new_node, Config) -> Config;
+init_per_testcase(new_node_accepts_long_old_chain_from_other_new_node_up_to_height_of_new_protocol, Config) -> Config;
+init_per_testcase(new_node_can_receive_short_old_chain_from_other_new_node, Config) -> Config;
+init_per_testcase(new_node_can_mine_on_old_chain_using_old_protocol, Config) -> Config;
+init_per_testcase(new_node_can_mine_on_old_chain_using_new_protocol, Config) -> Config.
+
+end_per_testcase(old_node_persisting_chain_and_not_mining_has_genesis_as_top, Config) ->
+    aest_nodes:ct_cleanup(Config);
+end_per_testcase(new_node_persisting_chain_and_not_mining_has_same_old_genesis_as_top, Config) ->
+    aest_nodes:ct_cleanup(Config);
+end_per_testcase(new_nodes_can_mine_and_sync_fast_minimal_chain_with_pow, Config) ->
+    aest_nodes:ct_cleanup(Config);
+end_per_testcase(restore_db_backup_on_old_node, Config) ->
+    aest_nodes:ct_cleanup(Config);
+end_per_testcase(old_node_can_receive_chain_from_other_old_node, Config) ->
+    aest_nodes:ct_cleanup(Config);
+end_per_testcase(restore_db_backup_with_short_chain_on_new_node, _) -> ok;
+end_per_testcase(new_node_accepts_long_old_chain_from_other_new_node_up_to_height_of_new_protocol, _) -> ok;
+end_per_testcase(new_node_can_receive_short_old_chain_from_other_new_node, _) -> ok;
+end_per_testcase(new_node_can_mine_on_old_chain_using_old_protocol, _) -> ok;
+end_per_testcase(new_node_can_mine_on_old_chain_using_new_protocol, _) -> ok.
```

Comment:
Lot's of duplication here, can we find a better way to distribute the `ct_setup` and `ct_cleanup` calls?

Reply:
You are correct - this shall be a macro `?is_in_group_upgrade_flow_smoke_test(TC)`. Though as cosmetic I plan not to address in time for this release. May address in follow up.

Reply:
Shouldn't groups be used for exactly this case?